### ISSUE TITLE
large_metadata_bad_client_test flake due to server validator not completing

### ISSUE THREAD

**Issue by markdroth** [grpc/grpc#11745]
This has been seen on Mac and Windows:

```
E0707 20:12:51.131000000 57308 bad_client.c:150] assertion failed: gpr_event_wait(&a.done_thd, grpc_timeout_seconds_to_deadline(5))
```

**Comment by jtattermusch** [comment]
Started seeing this with pretty high frequency on linux too:
https://sponge.corp.google.com/target?id=b703fb1d-a103-464a-ba4f-8722c11177b4&target=github/grpc/c_linux_dbg_native&searchFor=&show=FAILED&sortBy=STATUS

**Comment by markdroth** [comment]
I've finally found time to look into this, and it looks like it's a timing problem.  Basically, the chttp2 code uses the _acked_ settings to impose limits on what it will accept from the client -- if the server has sent a settings frame setting `SETTINGS_MAX_HEADER_LIST_SIZE` to 8KiB but the client has not yet acked those settings, then the server will not enforce them yet.  This is correct behavior for HTTP/2, but it causes problems for our test.

What our test is doing is sending a whole bunch of HTTP/2 data, including the settings frame, acking the server's settings frame, and sending a request with too much metadata, all in a single write.  In most runs of this test, the server sends its own settings frame and acks the client's settings frame before reading the client's request.  However, sometimes the server reads everything the client has sent before sending its own settings frame, which means that it does not consider the client's ack to have acknowledged its own settings frame by the time it receives the request with too much metadata, with the result that it does not impose the right metadata limit in the test.

This is unfortunately not trivial to fix.  @ctiller suggests that the best thing to do would be to use a more sophisticated framework that understands more of the HTTP/2 semantics.  One possible way of doing this will be to add a submodule of an OSS HTTP/2 library (maybe nghttp2) that we can use in tests, so that we can have more understanding of the communication with the server.  That will allow us to change write a client that will wait to receive the server's settings frame, send an ack, and then send a request with too much metadata, thus avoiding the flakiness of the current test.  It will also allow us to write some other, better tests in the future.

Yash has agreed to take this on, but he's got some other work in his queue ahead of this, so it's not clear when this will actually get done.

**Comment by jtattermusch** [comment]
Not seen in >2months.

**Comment by markdroth** [comment]
This hasn't been seen because Yash temporarily disabled the test until we have time to fix it.  It still does need to be fixed.

**Comment by yashykt** [comment]
Fixed in #14112
